src/client/oicp/CpoOICPClient.ts

```diff
@@ -0,0 +1,1272 @@
+    payload.Identification = transaction.oicpData.session.identification;
+
+    // Log
+    Logging.logDebug({
+      tenantID: this.tenant.id,
+      action: ServerAction.OICP_AUTHORIZE_STOP,
+      message: 'Stop Authorization',
+      module: MODULE_NAME, method: 'authorizeStop',
+      detailedMessages: { payload }
+    });
+
+    // Call Hubject
+    await this.axiosInstance.post(fullUrl, payload
+    ).then(function(response) {
+      console.log('Success! authorizeStop: ', response);
+      authorizeResponse = response.data as OICPAuthorizeStopCpoReceive;
+    }).catch((error) => {
+      console.log('Error! authorizeStop: ', error);
+    });
+
+    return authorizeResponse;
+  }
+
+  /**
+   * ERoaming Push Charge Detail Record
+   */
+  async pushCdr(transaction: Transaction): Promise<OICPAcknowledgment> {
+    let pushCdrResponse = {} as OICPAcknowledgment;
+    if (!transaction.oicpData) {
+      throw new BackendError({
+        source: transaction.chargeBoxID,
+        action: ServerAction.OICP_PUSH_CDRS,
+        message: `OICP data does not exists on Session ID '${transaction.id}'`,
+        module: MODULE_NAME, method: 'pushCdr',
+      });
+    }
+    if (!transaction.oicpData.session) {
+      throw new BackendError({
+        source: transaction.chargeBoxID,
+        action: ServerAction.OICP_PUSH_CDRS,
+        message: `OICP Session data does not exists on Session ID '${transaction.id}'`,
+        module: MODULE_NAME, method: 'pushCdr',
+      });
+    }
+    if (!transaction.stop) {
+      throw new BackendError({
+        source: transaction.chargeBoxID,
+        action: ServerAction.OICP_PUSH_CDRS,
+        message: `OICP Session ID '${transaction.oicpData.session.id}' (ID '${transaction.id}') not stopped`,
+        module: MODULE_NAME, method: 'pushCdr',
+      });
+    }
+
+    // Get CDR endpoint url
+    // for QA - environment: https://service-qa.hubject.com/api/oicp/cdrmgmt/v22/operators/{operatorID}/charge-detail-record
+    // for PROD - environment: https://service.hubject.com/api/oicp/cdrmgmt/v22/operators/{operatorID}/charge-detail-record
+    const fullUrl = this.getEndpointUrl('cdrs', ServerAction.OICP_PUSH_CDRS);
+
+    const cdr: OICPChargeDetailRecord = {} as OICPChargeDetailRecord;
+    cdr.SessionID = transaction.oicpData.session.id;
+    if (transaction.id) {
+      cdr.CPOPartnerSessionID = String(transaction.id); // Optional
+    }
+    cdr.EMPPartnerSessionID = transaction.oicpData.session.empPartnerSessionID; // Optional
+    cdr.EvseID = transaction.oicpData.session.evse.EvseID;
+    cdr.Identification = transaction.oicpData.session.identification;
+    cdr.ChargingStart = transaction.timestamp;
+    cdr.ChargingEnd = transaction.stop.timestamp;
+    cdr.SessionStart = transaction.oicpData.session.start_datetime;
+    cdr.SessionEnd = transaction.oicpData.session.end_datetime;
+    cdr.MeterValueStart = parseFloat((transaction.meterStart / 1000).toFixed(3)); // Optional
```

Comment:
Please use Utils.ConvertToFloat() and Utils.roundTo() helpers.

Reply:
ok, done